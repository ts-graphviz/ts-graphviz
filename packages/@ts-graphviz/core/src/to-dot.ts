import { DotObjectModel } from '@ts-graphviz/common';
import { stringify, fromModel, ConvertFromModelOptions, PrintOptions } from '@ts-graphviz/ast';

/**
 * This interface provides options for converting a model to DOT.
 * @group Convert Model to DOT
 */
export interface ToDotOptions {
  /**
   * Options for converting the model to DOT.
   */
  convert?: ConvertFromModelOptions;
  /**
   * Options for printing DOT.
   */
  print?: PrintOptions;
}

/**
 * Convert Model to DOT string.
 *
 * @group Convert Model to DOT
 *
 * @param model Dot Object Model, like {@link Digraph}, {@link Graph}, {@link Subgraph}, {@link Node}, and {@link Edge}
 * @param options Optional options for the conversion.
 * @returns DOT string
 */
export function toDot(model: DotObjectModel, options?: ToDotOptions): string {
  const ast = fromModel(model, options?.convert);
  return stringify(ast, options?.print);
}
