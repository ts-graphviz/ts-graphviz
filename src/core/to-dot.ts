import { DotObjectModel } from '#lib/common';
import { stringify, fromModel, ConvertFromModelOptions, PrintOptions } from '#lib/ast';

/**
 * @group Convert Model to DOT
 * @alpha
 */
export interface ToDotOptions {
  convert?: ConvertFromModelOptions;
  print?: PrintOptions;
}

/**
 * Convert Model to DOT string.
 *
 * @group Convert Model to DOT
 *
 * @param model Dot Object Model, like {@link Digraph}, {@link Graph}, {@link Node}, and {@link Edge}
 * @param options
 * @returns DOT string
 */
export function toDot(model: DotObjectModel, options?: ToDotOptions): string {
  const ast = fromModel(model, options?.convert);
  return stringify(ast, options?.print);
}
