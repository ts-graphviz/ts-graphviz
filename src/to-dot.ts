import { DotObjectModel } from './common/index.js';
import { stringify, fromModel, PrintOptions } from './ast/index.js';
import { ConvertOptions } from './ast/index.js';

export interface ToDotOptions {
  convert?: ConvertOptions;
  print?: PrintOptions;
}

export function toDot(model: DotObjectModel, options?: ToDotOptions): string {
  const ast = fromModel(model, options?.convert);
  return stringify(ast, options?.print);
}
