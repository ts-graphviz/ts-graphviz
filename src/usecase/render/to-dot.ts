import { Renderer, Dot } from './renderer.js';

export function toDot(object: Dot): string {
  const renderer = new Renderer();
  return renderer.render(object);
}
