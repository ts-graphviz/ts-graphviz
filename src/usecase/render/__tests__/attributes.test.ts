import { Attributes } from '../../../model/index.js';
import { toDot } from '../to-dot.js';

describe('Attributes rendering', () => {
  let attrs: Attributes;
  beforeEach(() => {
    attrs = new Attributes();
  });

  describe('renders correctly by toDot method', () => {
    it('no attributes', () => {
      expect(toDot(attrs)).toMatchSnapshot();
    });

    it('one attribute', () => {
      attrs.set('label', 'test');
      expect(toDot(attrs)).toMatchSnapshot();
    });

    it('some attributes', () => {
      attrs.set('label', 'test');
      attrs.set('color', 'red');
      expect(toDot(attrs)).toMatchSnapshot();
    });

    test('set some attributes by apply', () => {
      attrs.apply({
        label: 'this is test',
        color: 'red',
        fontsize: 16,
      });
      expect(toDot(attrs)).toMatchSnapshot();
    });

    test('set undefined attributes by apply', () => {
      attrs.apply({
        label: undefined,
      });
      expect(attrs.size).toBe(0);
      expect(toDot(attrs)).toMatchSnapshot();
    });

    describe('edge with comment', () => {
      beforeEach(() => {
        attrs.set('label', 'test');
        attrs.set('color', 'red');
      });
      test('single line comment', () => {
        attrs.comment = 'this is comment.';
        expect(toDot(attrs)).toMatchSnapshot();
      });

      test('multi line comment', () => {
        attrs.comment = 'this is comment.\nsecond line.';
        expect(toDot(attrs)).toMatchSnapshot();
      });
    });
  });
});
