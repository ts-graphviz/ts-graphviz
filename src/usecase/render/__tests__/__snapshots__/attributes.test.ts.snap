// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Attributes rendering renders correctly by toDot method edge with comment multi line comment 1`] = `
"[
  // this is comment.
  // second line.
  label = \\"test\\",
  color = \\"red\\",
]"
`;

exports[`Attributes rendering renders correctly by toDot method edge with comment single line comment 1`] = `
"[
  // this is comment.
  label = \\"test\\",
  color = \\"red\\",
]"
`;

exports[`Attributes rendering renders correctly by toDot method no attributes 1`] = `""`;

exports[`Attributes rendering renders correctly by toDot method one attribute 1`] = `
"[
  label = \\"test\\",
]"
`;

exports[`Attributes rendering renders correctly by toDot method set some attributes by apply 1`] = `
"[
  label = \\"this is test\\",
  color = \\"red\\",
  fontsize = 16,
]"
`;

exports[`Attributes rendering renders correctly by toDot method set undefined attributes by apply 1`] = `""`;

exports[`Attributes rendering renders correctly by toDot method some attributes 1`] = `
"[
  label = \\"test\\",
  color = \\"red\\",
]"
`;
